x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.ts

```diff
@@ -211,20 +211,22 @@ export class ObservabilityAIAssistantClient {
 
         const next = async (nextMessages: Message[]): Promise<void> => {
           const lastMessage = last(nextMessages);
-
           const isUserMessage = lastMessage?.message.role === MessageRole.User;
 
-          const isUserMessageWithoutFunctionResponse = isUserMessage && !lastMessage?.message.name;
-
-          const contextFirst =
-            isUserMessageWithoutFunctionResponse && functionClient.hasFunction('context');
+          const indexOfLastUserMessage = findLastIndex(
+            nextMessages,
+            ({ message }) => message.role === MessageRole.User && !message.name
+          );
 
-          const isAssistantMessageWithFunctionRequest =
-            lastMessage?.message.role === MessageRole.Assistant &&
-            !!lastMessage?.message.function_call?.name;
+          const shouldInjectContext =
+            indexOfLastUserMessage !== -1 &&
+            nextMessages
+              .slice(indexOfLastUserMessage)
+              .every(({ message }) => message.function_call?.name !== 'context') &&
+            functionClient.hasFunction('context');
```

Comment:
really small nit, but maybe we can move this up as it's probably the most important one and feels a little hidden now

Reply:
Do you mean to the top of the `next` function? It already is as high as it can be. It uses the variables declared before it.